plugins/experimental/wasm/wasm_main.cc

```diff
@@ -46,6 +46,249 @@ struct WasmInstanceConfig {
+  } else {
+    towrite = 0;
+  }
+
+  do {
+    TSDebug(WASM_DEBUG_TAG, "[%s] inside transform handler loop", __FUNCTION__);
+    proxy_wasm::FilterDataStatus status = proxy_wasm::FilterDataStatus::Continue;
+
+    if (towrite == 0 && !empty_input) {
+      break;
+    }
+
+    TSDebug(WASM_DEBUG_TAG, "[%s] retrieving text and calling the wasm handler function", __FUNCTION__);
+    if (!empty_input) {
+      blk   = TSIOBufferReaderStart(ti->reserved_reader);
+      start = TSIOBufferBlockReadStart(blk, ti->reserved_reader, &blk_len);
+
+      int size = 0;
+      if (towrite > blk_len) {
+        c->setTransformResult(start, blk_len);
+        towrite -= blk_len;
+        TSIOBufferReaderConsume(ti->reserved_reader, blk_len);
+        size = blk_len;
+      } else {
+        c->setTransformResult(start, towrite);
+        TSIOBufferReaderConsume(ti->reserved_reader, towrite);
+        size    = towrite;
+        towrite = 0;
+      }
+
+      if (!towrite && eos) {
+        if (ti->request) {
+          status = c->onRequestBody(size, true);
+        } else {
+          status = c->onResponseBody(size, true);
+        }
+      } else {
+        if (ti->request) {
+          status = c->onRequestBody(size, false);
+        } else {
+          status = c->onResponseBody(size, false);
+        }
+      }
+    } else {
+      c->setTransformResult(nullptr, 0);
+      if (ti->request) {
+        status = c->onRequestBody(0, true);
+      } else {
+        status = c->onResponseBody(0, true);
+      }
+    }
+
+    TSDebug(WASM_DEBUG_TAG, "[%s] retrieving returns from wasm handler function and pass back to ATS", __FUNCTION__);
+    if ((status == proxy_wasm::FilterDataStatus::Continue) ||
+        ((status == proxy_wasm::FilterDataStatus::StopIterationAndBuffer ||
+          status == proxy_wasm::FilterDataStatus::StopIterationAndWatermark) &&
+         eos && !towrite)) {
+      res = c->getTransformResult(&res_len);
+
+      if (res && res_len > 0) {
+        if (!ti->output_vio) {
+          if (eos && !towrite) {
+            ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, res_len); // HttpSM go on
+          } else {
+            ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, ti->downstream_bytes); // HttpSM go on
+          }
+        }
+
+        TSIOBufferWrite(ti->output_buffer, res, res_len);
+        ti->total  += res_len;
+        write_down  = true;
+      }
+
+      c->clearTransformResult();
+    }
+
+    if (status == proxy_wasm::FilterDataStatus::StopIterationNoBuffer) {
+      c->clearTransformResult();
+    }
+
+    if (eos && !towrite) { // EOS
+      break;
+    }
+
+  } while (towrite > 0);
+
+  if (eos && !ti->output_vio) {
+    ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, 0);
+  }
+
+  if (write_down || eos) {
+    TSVIOReenable(ti->output_vio);
+  }
+
+  if (toread > input_avail) { // upstream not finished.
+    if (eos) {
```

Comment:
is this even possible? if `toread > input_avail`, `eos` will always be false since there is more to read? also from how `eos` is set above?

Reply:
it can be true by the function in the wasm module. I forgot to retrieve that value and set eos . 

Good find. So that's why you can still have the EOS event below.

Reply:
I just double checked the proxy-wasm spec. It currently does not allow changing the eos. But I expect it be possible in the future. So I want to keep the code as if. But let me add a bunch of comments to explain the idea